### DISCUSSION TITLE
FSP server support for loading GameCube games over the network

### DISCUSSION THREAD

**Discussion by fhoedemakers** [retronas/discussions#105] [Ideas]
Hi,

It would be a great feature to add [FSP server](http://fsp.sourceforge.net/) support, so GameCube games can be loaded over the Network from the NAS via [Swiss](https://github.com/emukidid/swiss-gc).

Frank

**Comment by danmons** [comment]
Aiming to look at this later this week, time permitting.  I don't have a GameCube Broadband Adaptor to test, so once it's committed I'll need volunteers to test.

**Comment by fhoedemakers** [comment]
> Aiming to look at this later this week, time permitting. I don't have a GameCube Broadband Adaptor to test, so once it's committed I'll need volunteers to test.

Sure, i can do some testing.

**Comment by sephallen** [comment]
I dockerised FSP server last year and currently run it on unRAID to play GameCube games via Swiss using a BBA.
You can find my repo here: https://github.com/sephallen/fsp-server-docker - hopefully it can help a little!

**Comment by fhoedemakers** [comment]
I also created a docker container for FSP. 

- One for intel https://hub.docker.com/r/frenskefrens/fsp-server which works great for loading GameCube  games via Docker on my Synology NAS.
- the other for ARM which i couldn't get to work properly for loading GameCube games via Docker on my Raspberry PI:  https://hub.docker.com/r/frenskefrens/fsp-server-arm

**Comment by sephallen** [comment]
I'm not sure if an ARM build of FSP server exists, though it might be able to be built from source.

**Comment by fhoedemakers** [comment]
> I'm not sure if an ARM build of FSP server exists, though it might be able to be built from source.

It is buildable from source for the Raspberry PI

**Comment by fhoedemakers** [comment]
> Aiming to look at this later this week, time permitting. I don't have a GameCube Broadband Adaptor to test, so once it's committed I'll need volunteers to test.
@danmons I can help with testing. I can also provide the necessary ARM binaries and configuration file setting.

**Comment by danmons** [comment]
I've had the ansible playbook for source building for about a week already, that's fine (was waiting for my BBA to turn up).  Tested and working with a command line FSP client from other machines.  I'll stick with source build because it's fast enough (less than a minute on an RPi4) and then it's cross-platform (userbase right now is a mix of ARM 32 and 64bit, as well as x86 32 and 64bit). 

My loaner GC BBA turned up yesterday, and initial testing worked straight away:
https://twitter.com/_daemons/status/1491308998100144131

Committed a few more changes today to make the FSP server run as an unprivileged user on UDP/2121 as Swiss lets you choose a custom port.  FSP interestingly has no "default" port.  The FSP devs suggest UDP/21 (with FTP being TCP/21), but again that's not enforced or expected by anyone as far as I can see.  Daemons running as root should be reduced in number if we can, and FSP also does caching by dropping `.FSP_CONTENT` files all over the disk with information in them to speed up how it searches for files.  With the daemon running as root, even with config to tell it switch to the RetroNAS user ("pi" by default), it still writes these cache files as root and the standard user can't clean them up without running the permission fix tool. 

So far GameCube is the only system I know of that uses FSP.  With that in mind I'm making the FSP root the GameCube ISO directory (via a dedicated `gamecube/swiss/games` folder and a symlink into `roms/nitendo/gamecube/iso`.   If other systems ever decide to use FSP (which I think they should, it's a wonderfully light weight protocol), we can revisit this choice.  Potentially one daemon/port per device or something. But that's for a future date/time. 

I'll do one more test shortly of this new UDP/2121 config.  When that's confirmed working I'll push the menu items to let people install the tool and start testing.  ETA about 6 hours from now (have to take care of real life things first).

**Comment by Extrems** [comment]
To support future development, there should be a `swiss` directory under the server's home directory, and it should contain the following blank files:

- `.FSP_OK_ADD`
- `.FSP_OK_DEL`
- `.FSP_OK_MKDIR`
- `.FSP_OK_RENAME`

These may be in the home directory instead if security is not a concern, while leaving the creation of the `swiss` directory to Swiss.
